core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -97,6 +97,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         case requestId => throw new KafkaException("Unknown api code " + requestId)
       }
     } catch {
+      case fee: FatalExitError => throw fee
```

Comment:
Let me make sure I remember the deadlock correctly: a fetcher thread calls System.exit and waits until the shutdownhook thread finishes. The shutdownhook thread uses ShutdownableThread latch to wait on each thread including the fetcher thread. In this case the status of the latch in the fetcher thread is not changed to indicate that it has started the shut down process. So I am not sure at which point the check for the latch being open is performed to avoid the deadlock.